grafana/grafana
# loki: backend mode: support all query types

run all explore-queries through loki-backend-mode.

NOTE: later we will switch to use a streaming-json-parser, to reduce memory use, this is a step in that direction, so the code splits the  loki_json -> grafana_dataframes processing into two steps:
1. loki_json -> grafana_dataframes without any customization (this will be replaced later by different code)
2. adjust metadata like field-names etc. on the grafana dataframes to be like frontend&alerting wants them

how to test:
- make sure you have the loki-backend-mode enabled in your `conf/custom.ini` like this:
    ```yaml
    [feature_toggles]
    lokiBackendMode = true
    ```
- make sure you have a running loki datasource
- go to explore-mode, and run some queries, like:
    - metric: `count_over_time({level="info"}[10s)`, try both instant and ranged
    - logs: `{level="info"}`, (i recommend trying only ranged, instant-logs-queries are a problematic concept :) )
    - verify in the browser-developer-tools that the ajax requests are going to `/api/ds/query`, not to `/api/datasources/proxy/...`
        - note, loki-queries that load metadata, like list-of-tags for the autocomplete-system are still running in frontend-mode, so they would hit the proxy-endpoint
 - verify that log-volume-queries set the custom http header
     - when running the log-volume queries, we have to set the `X-Query-Tags: Source=logvolhist` header.
     - to test it:
         - you will need a way to inspect the HTTP requests traveling between grafana-go-server and Loki. a proxy is a good option, i recommend [mitmproxy](https://mitmproxy.org). if you need help with this step, ping @gabor 😄 
         - (make sure you are in RANGE mode, not INSTANT)
         - run a log-producing query, like `{level="info"}`
         - you should see two requests going from grafana-go-server to Loki
           - the logs-query that you specified
           - another query like `sum by (level) (count_over_time(<your_original_query>...))` that calculates the log-volume-graph. this http request must have the http header `X-Query-Tags` set to `Source=logvolhist`
